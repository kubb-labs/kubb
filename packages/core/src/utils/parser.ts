import { extname } from 'node:path'
import type * as KubbFile from '@kubb/fs/types'

import { getRelativePath } from '@kubb/fs'
import hash from 'object-hash'
import { combineExports, combineImports, combineSources } from '../FileManager.ts'
import type { Logger } from '../logger.ts'

/**
 * Helper to create a file with name and id set
 */
export function createFile<TMeta extends object = object>(file: KubbFile.File<TMeta>): KubbFile.ResolvedFile<TMeta> {
  const extName = extname(file.baseName) as KubbFile.Extname

  if (!extName) {
    throw new Error(`No extName found for ${file.baseName}`)
  }

  const source = file.sources.map((item) => item.value).join('\n\n')
  const exports = file.exports ? combineExports(file.exports) : []
  const imports = file.imports && source ? combineImports(file.imports, exports, source) : []
  const sources = file.sources ? combineSources(file.sources) : []

  return {
    ...file,
    id: hash({ path: file.path }),
    name: trimExtName(file.baseName),
    extName,
    imports: imports.map((item) => createFileImport(item)),
    exports: exports.map((item) => createFileExport(item)),
    sources: sources.map((item) => createFileSource(item)),
  }
}

/**
 * Helper to create a fileImport with extName set
 */
export function createFileSource(source: KubbFile.Source): KubbFile.Source {
  return source
}

/**
 * Helper to create a fileImport with extName set
 */
export function createFileImport(imp: KubbFile.Import): KubbFile.ResolvedImport {
  const extName = extname(imp.path) as KubbFile.Extname

  return {
    ...imp,
    extName: imp.extName ? imp.extName : extName,
  }
}

/**
 * Helper to create a fileExport with extName set
 */
export function createFileExport(exp: KubbFile.Export): KubbFile.ResolvedExport {
  const extName = extname(exp.path) as KubbFile.Extname

  return {
    ...exp,
    extName: exp.extName ? exp.extName : extName,
  }
}

export type ParserModule<TMeta extends object = object> = {
  /**
   * By default @kubb/react is used
   */
  render: (item: any) => any
  /**
   * Convert a file to string
   */
  print: (file: KubbFile.ResolvedFile<TMeta>, options: PrintOptions) => Promise<string>
}

export function createFileParser<TMeta extends object = object>(parser: ParserModule<TMeta>): ParserModule<TMeta> {
  return parser
}

type PrintOptions = {
  logger?: Logger
}

const typeScriptParser = createFileParser({
  render() {
    return undefined
  },
  async print(file) {
    const module = await import('@kubb/parser-ts')

    const source = file.sources.map((item) => item.value).join('\n\n')

    const importNodes = file.imports
      .map((item) => {
        const path = item.root ? getRelativePath(item.root, item.path) : item.path

        return module.factory.createImportDeclaration({
          name: item.name,
          path,
          isTypeOnly: item.isTypeOnly,
        })
      })
      .filter(Boolean)

    const exportNodes = file.exports
      .map((item) => {
        return module.factory.createExportDeclaration({
          name: item.name,
          path: item.path,
          isTypeOnly: item.isTypeOnly,
          asAlias: item.asAlias,
        })
      })
      .filter(Boolean)

    // do some basic linting with the ts compiler
    return module.print([], { source: [module.print([...importNodes, ...exportNodes]), source].join('\n'), noEmitHelpers: false })
  },
})

const defaultParser = createFileParser({
  render() {
    return undefined
  },
  async print(file, { logger }) {
    return file.sources.map((item) => item.value).join('\n\n')
  },
})

const parsers: Record<KubbFile.Extname, ParserModule<any>> = {
  '.ts': typeScriptParser,
  '.js': typeScriptParser,
  '.jsx': typeScriptParser,
  '.tsx': typeScriptParser,
  '.json': defaultParser,
}

export async function getFileParser<TMeta extends object = object>(extName: KubbFile.Extname | undefined): Promise<ParserModule<TMeta>> {
  if (!extName) {
    return defaultParser
  }

  const parser = parsers[extName]

  if (!parser) {
    console.warn(`[parser] No parser found for ${extName}, default parser will be used`)
  }

  return parser || defaultParser
}

export function trimExtName(text: string): string {
  const extName = text.split('.').pop()

  return text.replace(`.${extName}`, '')
}
