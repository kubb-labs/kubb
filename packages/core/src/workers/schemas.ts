import { z } from 'zod'

import type { ZodStringDef, ZodType } from 'zod'
import type { KubbFile } from '../FileManager.ts'

export const importSchema = z.object({
  name: z.union([z.string(), z.array(z.string())]),
  path: z.string(),
  isTypeOnly: z.boolean().optional(),
  isNameSpace: z.boolean().optional(),
  root: z.string().optional(),
})

export const exportSchema = z.object({
  name: z.union([z.string(), z.array(z.string())]).optional(),
  path: z.string(),
  isTypeOnly: z.boolean().optional(),
  asAlias: z.boolean().optional(),
})

export const fileSchema = z.object({
  id: z.string().optional(),
  baseName: z.string() as ZodType<KubbFile.BaseName, ZodStringDef>,
  path: z.string(),
  source: z.string(),
  imports: z.array(importSchema).optional(),
  exports: z.array(exportSchema).optional(),
  override: z.boolean().optional(),
  meta: z.object({}).optional(),
  env: z.object({}).optional(),
})
