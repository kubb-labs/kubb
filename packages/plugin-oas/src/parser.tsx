import type { PluginFactoryOptions } from '@kubb/core'
import type { Operation, SchemaObject } from '@kubb/oas'
import { App, createRoot, type KubbNode } from '@kubb/react'
import type * as KubbFile from '@kubb/fs/types'
import type { OperationsByMethod } from './types.ts'
import { Oas } from '@kubb/plugin-oas/components'
import type { OperationGenerator } from './OperationGenerator.ts'
import type { SchemaGenerator, SchemaGeneratorOptions } from './SchemaGenerator.ts'

type OperationsProps<TOptions extends PluginFactoryOptions> = {
  generator: OperationGenerator<TOptions>
  options: TOptions['resolvedOptions']
  operations: Array<Operation>
  operationsByMethod: OperationsByMethod
}

type OperationProps<TOptions extends PluginFactoryOptions> = {
  generator: OperationGenerator<TOptions>
  options: TOptions['resolvedOptions']
  operation: Operation
}

type SchemaProps<TOptions extends PluginFactoryOptions> = {
  generator: SchemaGenerator<SchemaGeneratorOptions, TOptions>
  name: string
  schema: SchemaObject
  options: TOptions['resolvedOptions']
}

export type ParserOptions<TOptions extends PluginFactoryOptions> = {
  name: string
  operations?: (props: OperationsProps<TOptions>) => Promise<KubbFile.File[]>
  operation?: (props: OperationProps<TOptions>) => Promise<KubbFile.File[]>
  schema?: (props: SchemaProps<TOptions>) => Promise<KubbFile.File[]>
}

export type Parser<TOptions extends PluginFactoryOptions> = ParserOptions<TOptions>

export function createParser<TOptions extends PluginFactoryOptions>(parseOptions: ParserOptions<TOptions>): Parser<TOptions> {
  return parseOptions
}

export type ParserReactOptions<TOptions extends PluginFactoryOptions> = {
  name: string
  Operations?: (props: OperationsProps<TOptions>) => KubbNode
  Operation?: (props: OperationProps<TOptions>) => KubbNode
  Schema?: (props: SchemaProps<TOptions>) => KubbNode
}

export function createReactParser<TOptions extends PluginFactoryOptions>(parseOptions: ParserReactOptions<TOptions>): Parser<TOptions> {
  return {
    ...parseOptions,
    async operations({ generator, options, operations, operationsByMethod }) {
      if (!parseOptions.Operations) {
        return []
      }

      const { pluginManager, oas, plugin, mode } = generator.context
      const root = createRoot({
        logger: pluginManager.logger,
      })

      root.render(
        <App pluginManager={pluginManager} plugin={plugin} mode={mode}>
          <Oas oas={oas} operations={operations} generator={generator}>
            <parseOptions.Operations operations={operations} generator={generator} operationsByMethod={operationsByMethod} options={options} />
          </Oas>
        </App>,
      )

      return root.files
    },
    async operation({ generator, operation, options }) {
      if (!parseOptions.Operation) {
        return []
      }

      const { pluginManager, oas, plugin, mode } = generator.context
      const root = createRoot({
        logger: pluginManager.logger,
      })

      root.render(
        <App pluginManager={pluginManager} plugin={{ ...plugin, options }} mode={mode}>
          <Oas oas={oas} operations={[operation]} generator={generator}>
            <Oas.Operation operation={operation}>
              <parseOptions.Operation operation={operation} options={options} generator={generator} />
            </Oas.Operation>
          </Oas>
        </App>,
      )

      return root.files
    },
    async schema({ generator, schema, name, options }) {
      if (!parseOptions.Schema) {
        return []
      }

      const { pluginManager, oas, plugin, mode } = generator.context
      const root = createRoot({
        logger: pluginManager.logger,
      })

      const tree = generator.parse({ schema, name })

      root.render(
        <App pluginManager={pluginManager} plugin={{ ...plugin, options }} mode={mode}>
          <Oas oas={oas}>
            <Oas.Schema name={name} value={schema} tree={tree}>
              <parseOptions.Schema schema={schema} options={options} generator={generator} name={name} />
            </Oas.Schema>
          </Oas>
        </App>,
      )

      return root.files
    },
  }
}
