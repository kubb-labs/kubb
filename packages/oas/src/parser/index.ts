import OASNormalize from 'oas-normalize'
import swagger2openapi from 'swagger2openapi'

import { Oas } from '../Oas.ts'
import { filterAndSort, isOpenApiV2Document } from '../utils.ts'

import type { OASDocument } from 'oas/types'
import type { OpenAPI } from 'openapi-types'

export type FormatOptions = {
  verbose?: boolean
  'no-sort'?: boolean
  sort?: boolean
  output?: string
  sortSet?: {
    root?: Array<'openapi' | 'info' | 'servers' | 'paths' | 'components' | 'tags' | 'x-tagGroups' | 'externalDocs'>
    get?: Array<'operationId' | 'summary' | 'description' | 'parameters' | 'requestBody' | 'responses'>
    post?: Array<'operationId' | 'summary' | 'description' | 'parameters' | 'requestBody' | 'responses'>
    put?: Array<'operationId' | 'summary' | 'description' | 'parameters' | 'requestBody' | 'responses'>
    patch?: Array<'operationId' | 'summary' | 'description' | 'parameters' | 'requestBody' | 'responses'>
    delete?: Array<'operationId' | 'summary' | 'description' | 'parameters' | 'requestBody' | 'responses'>
    parameters?: Array<'name' | 'in' | 'description' | 'required' | 'schema'>
    requestBody?: Array<'description' | 'required' | 'content'>
    responses?: Array<'description' | 'headers' | 'content' | 'links'>
    content?: Array<string>
    components?: Array<'parameters' | 'schemas'>
    schema?: Array<'description' | 'type' | 'items' | 'properties' | 'format' | 'example' | 'default'>
    schemas?: Array<'description' | 'type' | 'items' | 'properties' | 'format' | 'example' | 'default'>
    properties?: Array<'description' | 'type' | 'items' | 'format' | 'example' | 'default' | 'enum'>
  }
  filterSet?: {
    methods?: Array<'get' | 'post' | 'put' | 'patch' | 'delete' | 'options' | 'trace' | 'head' | 'parameters'>
    inverseMethods?: Array<'get' | 'post' | 'put' | 'patch' | 'delete' | 'options' | 'trace' | 'head' | 'parameters'>
    tags?: Array<string>
    inverseTags?: Array<string>
    operationIds?: Array<string>
    inverseOperationIds?: Array<string>
    operations?: Array<string>
    flags?: Array<string>
    inverseFlags?: Array<string>
    flagValues?: Array<string>
    inverseFlagValues?: Array<string>
    stripFlags?: Array<string>
    responseContent?: Array<string>
    inverseResponseContent?: Array<string>
    unusedComponents?: Array<'requestBodies' | 'schemas' | 'parameters' | 'responses'>
  }
  sortComponentsSet?: {}
  casingSet?: {}
}

export async function parse(pathOrApi: string | OASDocument, options: FormatOptions = {}): Promise<Oas> {
  const oasNormalize = new OASNormalize(pathOrApi, {
    enablePaths: true,
    colorizeErrors: true,
  })
  let document: OpenAPI.Document

  try {
    // resolve external refs
    document = await oasNormalize.bundle()
  } catch (e) {
    document = (await oasNormalize.load()) as OpenAPI.Document
  }

  if (isOpenApiV2Document(document)) {
    const { openapi } = await swagger2openapi.convertObj(document, {
      anchors: true,
    })

    const oas = await filterAndSort(openapi as OASDocument, options)

    return new Oas({ oas: oas as OASDocument })
  }

  const oas = await filterAndSort(document as OASDocument, options)

  return new Oas({ oas })
}
