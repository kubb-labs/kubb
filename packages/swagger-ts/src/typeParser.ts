import transformers from '@kubb/core/transformers'
import { print } from '@kubb/parser'
import * as factory from '@kubb/parser/factory'
import { isKeyword, SchemaGenerator, schemaKeywords } from '@kubb/swagger'

import type { ts } from '@kubb/parser'
import type { Schema, SchemaKeywordMapper, SchemaMapper } from '@kubb/swagger'

export const typeKeywordMapper = {
  any: () => factory.keywordTypeNodes.any,
  unknown: () => factory.keywordTypeNodes.unknown,
  number: () => factory.keywordTypeNodes.number,
  integer: () => factory.keywordTypeNodes.number,
  object: (nodes?: ts.TypeElement[]) => {
    if (!nodes) {
      return factory.keywordTypeNodes.object
    }

    return factory.createTypeLiteralNode(nodes)
  },
  lazy: undefined,
  string: () => factory.keywordTypeNodes.string,
  boolean: () => factory.keywordTypeNodes.boolean,
  undefined: () => factory.keywordTypeNodes.undefined,
  nullable: undefined,
  null: () => factory.keywordTypeNodes.null,
  nullish: undefined,
  array: (nodes?: ts.TypeNode[]) => {
    if (!nodes) {
      return undefined
    }

    return factory.createArrayDeclaration({ nodes })
  },
  tuple: (nodes?: ts.TypeNode[]) => {
    if (!nodes) {
      return undefined
    }

    return factory.createTupleTypeNode(nodes)
  },
  enum: (name?: string) => {
    if (!name) {
      return undefined
    }

    return factory.createTypeReferenceNode(name, undefined)
  },
  union: (nodes?: ts.TypeNode[]) => {
    if (!nodes) {
      return undefined
    }

    return factory.createUnionDeclaration({
      withParentheses: true,
      nodes,
    })
  },
  const: (name?: string | number, format?: 'string' | 'number') => {
    if (!name) {
      return undefined
    }

    if (format === 'number') {
      return factory.createLiteralTypeNode(factory.createNumericLiteral(name))
    }

    return factory.createLiteralTypeNode(factory.createStringLiteral(name.toString()))
  },
  datetime: () => factory.keywordTypeNodes.string,
  date: () => factory.createTypeReferenceNode(factory.createIdentifier('Date')),
  uuid: undefined,
  url: undefined,
  strict: undefined,
  default: undefined,
  and: (nodes?: ts.TypeNode[]) => {
    if (!nodes) {
      return undefined
    }

    return factory.createIntersectionDeclaration({
      withParentheses: true,
      nodes,
    })
  },
  describe: undefined,
  min: undefined,
  max: undefined,
  optional: undefined,
  matches: undefined,
  email: undefined,
  firstName: undefined,
  lastName: undefined,
  password: undefined,
  phone: undefined,
  readOnly: undefined,
  ref: (propertyName?: string) => {
    if (!propertyName) {
      return undefined
    }

    return factory.createTypeReferenceNode(propertyName, undefined)
  },
  blob: () => factory.createTypeReferenceNode('Blob', []),
  deprecated: undefined,
  example: undefined,
  type: undefined,
  format: undefined,
  catchall: undefined,
} satisfies SchemaMapper<(ctx?: any) => ts.Node | null | undefined>

type ParserOptions = {
  name: string
  typeName?: string
  description?: string
  /**
   * @default `'questionToken'`
   */
  optionalType: 'questionToken' | 'undefined' | 'questionTokenAndUndefined'
  /**
   * @default `'asConst'`
   */
  enumType: 'enum' | 'asConst' | 'asPascalConst' | 'constEnum' | 'literal'
  keysToOmit?: string[]
  mapper?: typeof typeKeywordMapper
}

export function parseTypeMeta(item: Schema, options: ParserOptions): ts.Node | null | undefined {
  const mapper = options.mapper || typeKeywordMapper
  const value = mapper[item.keyword as keyof typeof mapper]

  if (!value) {
    return undefined
  }

  if (isKeyword(item, schemaKeywords.union)) {
    const value = mapper[item.keyword as keyof typeof mapper] as (typeof typeKeywordMapper)['union']
    return value(item.args.map((orItem) => parseTypeMeta(orItem, options)).filter(Boolean) as ts.TypeNode[])
  }

  if (isKeyword(item, schemaKeywords.and)) {
    const value = mapper[item.keyword as keyof typeof mapper] as (typeof typeKeywordMapper)['and']
    return value(item.args.map((orItem) => parseTypeMeta(orItem, options)).filter(Boolean) as ts.TypeNode[])
  }

  if (isKeyword(item, schemaKeywords.array)) {
    const value = mapper[item.keyword as keyof typeof mapper] as (typeof typeKeywordMapper)['array']
    return value(item.args.items.map((orItem) => parseTypeMeta(orItem, options)).filter(Boolean) as ts.TypeNode[])
  }

  if (isKeyword(item, schemaKeywords.enum)) {
    const value = mapper[item.keyword as keyof typeof mapper] as (typeof typeKeywordMapper)['enum']
    return value(item.args.typeName)
  }

  if (isKeyword(item, schemaKeywords.ref)) {
    const value = mapper[item.keyword as keyof typeof mapper] as (typeof typeKeywordMapper)['ref']
    return value(item.args.name)
  }

  if (isKeyword(item, schemaKeywords.blob)) {
    return value()
  }

  if (isKeyword(item, schemaKeywords.tuple)) {
    const value = mapper[item.keyword as keyof typeof mapper] as (typeof typeKeywordMapper)['tuple']
    return value(item.args.map((tupleItem) => parseTypeMeta(tupleItem, options)).filter(Boolean) as ts.TypeNode[])
  }

  if (isKeyword(item, schemaKeywords.const)) {
    const value = mapper[item.keyword as keyof typeof mapper] as (typeof typeKeywordMapper)['const']

    return value(item.args.name, item.args.format)
  }

  if (isKeyword(item, schemaKeywords.object)) {
    const value = mapper[item.keyword as keyof typeof mapper] as (typeof typeKeywordMapper)['object']

    const properties = Object.entries(item.args?.properties || {})
      .filter((item) => {
        const schemas = item[1]
        return schemas && typeof schemas.map === 'function'
      })
      .map((item) => {
        const name = item[0]
        const schemas = item[1]

        const isNullish = schemas.some((item) => item.keyword === schemaKeywords.nullish)
        const isNullable = schemas.some((item) => item.keyword === schemaKeywords.nullable)
        const isOptional = schemas.some((item) => item.keyword === schemaKeywords.optional)
        const isReadonly = schemas.some((item) => item.keyword === schemaKeywords.readOnly)
        const describeSchema = schemas.find((item) => item.keyword === schemaKeywords.describe) as SchemaKeywordMapper['describe'] | undefined
        const deprecatedSchema = schemas.find((item) => item.keyword === schemaKeywords.deprecated) as SchemaKeywordMapper['deprecated'] | undefined
        const defaultSchema = schemas.find((item) => item.keyword === schemaKeywords.default) as SchemaKeywordMapper['default'] | undefined
        const exampleSchema = schemas.find((item) => item.keyword === schemaKeywords.example) as SchemaKeywordMapper['example'] | undefined
        const typeSchema = schemas.find((item) => item.keyword === schemaKeywords.type) as SchemaKeywordMapper['type'] | undefined
        const formatSchema = schemas.find((item) => item.keyword === schemaKeywords.format) as SchemaKeywordMapper['format'] | undefined

        let type = schemas.map((item) => parseTypeMeta(item, options)).filter(Boolean)[0] as ts.TypeNode

        if (isNullable) {
          type = factory.createUnionDeclaration({
            nodes: [type, factory.keywordTypeNodes.null],
          }) as ts.TypeNode
        }

        if (isNullish && ['undefined', 'questionTokenAndUndefined'].includes(options.optionalType as string)) {
          type = factory.createUnionDeclaration({
            nodes: [type, factory.keywordTypeNodes.undefined],
          }) as ts.TypeNode
        }

        if (isOptional && ['undefined', 'questionTokenAndUndefined'].includes(options.optionalType as string)) {
          type = factory.createUnionDeclaration({
            nodes: [type, factory.keywordTypeNodes.undefined],
          }) as ts.TypeNode
        }

        const propertySignature = factory.createPropertySignature({
          questionToken: isOptional && ['questionToken', 'questionTokenAndUndefined'].includes(options.optionalType as string),
          name,
          type,
          readOnly: isReadonly,
        })

        return factory.appendJSDocToNode({
          node: propertySignature,
          comments: [
            describeSchema ? `@description ${transformers.jsStringEscape(describeSchema.args)}` : undefined,
            deprecatedSchema ? '@deprecated' : undefined,
            defaultSchema ? `@default ${defaultSchema.args}` : undefined,
            exampleSchema ? `@example ${exampleSchema.args}` : undefined,
            typeSchema ? `@type ${typeSchema.args}${!isOptional ? '' : ' | undefined'} ${formatSchema?.args || ''}` : undefined,
          ].filter(Boolean),
        })
      })

    const additionalProperties = item.args?.additionalProperties?.length
      ? factory.createIndexSignature(
          item.args.additionalProperties
            .map((schema) => parseTypeMeta(schema, options))
            .filter(Boolean)
            .at(0) as ts.TypeNode,
        )
      : undefined

    return value([...properties, additionalProperties].filter(Boolean))
  }

  if (item.keyword in mapper) {
    return value()
  }

  return undefined
}

export function typeParser(schemas: Schema[], options: ParserOptions): string {
  const nodes: ts.Node[] = []
  const extraNodes: ts.Node[] = []

  if (!schemas.length) {
    return ''
  }

  const isNullish = schemas.some((item) => item.keyword === schemaKeywords.nullish)
  const isNullable = schemas.some((item) => item.keyword === schemaKeywords.nullable)
  const isOptional = schemas.some((item) => item.keyword === schemaKeywords.optional)

  let type =
    (schemas
      .map((schema) => parseTypeMeta(schema, options))
      .filter(Boolean)
      .at(0) as ts.TypeNode) || typeKeywordMapper.undefined()

  if (isNullable) {
    type = factory.createUnionDeclaration({
      nodes: [type, factory.keywordTypeNodes.null],
    }) as ts.TypeNode
  }

  if (isNullish && ['undefined', 'questionTokenAndUndefined'].includes(options.optionalType as string)) {
    type = factory.createUnionDeclaration({
      nodes: [type, factory.keywordTypeNodes.undefined],
    }) as ts.TypeNode
  }

  if (isOptional && ['undefined', 'questionTokenAndUndefined'].includes(options.optionalType as string)) {
    type = factory.createUnionDeclaration({
      nodes: [type, factory.keywordTypeNodes.undefined],
    }) as ts.TypeNode
  }

  const node = factory.createTypeAliasDeclaration({
    modifiers: [factory.modifiers.export],
    name: options.name,
    type: options.keysToOmit?.length
      ? factory.createOmitDeclaration({
          keys: options.keysToOmit,
          type,
          nonNullable: true,
        })
      : type,
  })

  const enumSchemas = SchemaGenerator.deepSearch(schemas, schemaKeywords.enum)
  if (enumSchemas) {
    enumSchemas.forEach((enumSchema) => {
      extraNodes.push(
        ...factory.createEnumDeclaration({
          name: transformers.camelCase(enumSchema.args.name),
          typeName: enumSchema.args.typeName,
          enums: enumSchema.args.items
            .map((item) => (item.value === undefined ? undefined : [transformers.trimQuotes(item.name?.toString()), item.value]))
            .filter(Boolean) as unknown as [string, string][],
          type: options.enumType,
        }),
      )
    })
  }

  nodes.push(
    factory.appendJSDocToNode({
      node,
      comments: [options.description ? `@description ${transformers.jsStringEscape(options.description)}` : undefined].filter(Boolean),
    }),
  )

  const filterdNodes = nodes.filter(
    (node: ts.Node) =>
      !extraNodes.some(
        (extraNode: ts.Node) => (extraNode as ts.TypeAliasDeclaration)?.name?.escapedText === (node as ts.TypeAliasDeclaration)?.name?.escapedText,
      ),
  )

  return print([...extraNodes, ...filterdNodes])
}
