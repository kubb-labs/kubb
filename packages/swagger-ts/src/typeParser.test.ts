import { print } from '@kubb/parser'

import { schemas } from '../../swagger/mocks/schemas.ts'
import { parseTypeMeta, typeParser } from './typeParser.ts'

describe('parseTypeMeta', () => {
  test.each(schemas.basic)('$name', ({ schema, name }) => {
    const parsed = parseTypeMeta(schema, {
      name,
      optionalType: 'questionToken',
      enumType: 'asConst',
    })
    const text = parsed ? print(parsed) : undefined

    expect(text).toMatchSnapshot()
  })
})

describe('typeParser', () => {
  test.each(schemas.full)('$name', ({ schema, name }) => {
    const text = typeParser(schema, {
      name,
      optionalType: 'questionToken',
      enumType: 'asConst',
    })
    expect(text).toMatchSnapshot()
  })
})
