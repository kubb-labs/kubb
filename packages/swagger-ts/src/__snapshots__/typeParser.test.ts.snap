// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parseTypeMeta > 'and' 1`] = `
"(string & number)
"
`;

exports[`parseTypeMeta > 'any' 1`] = `
"any
"
`;

exports[`parseTypeMeta > 'array' 1`] = `
"(number | string)[]
"
`;

exports[`parseTypeMeta > 'arrayAdvanced' 1`] = `
"number[]
"
`;

exports[`parseTypeMeta > 'arrayEmpty' 1`] = `
"[
]
"
`;

exports[`parseTypeMeta > 'arrayRef' 1`] = `
"Pet[]
"
`;

exports[`parseTypeMeta > 'blob' 1`] = `
"Blob
"
`;

exports[`parseTypeMeta > 'boolean' 1`] = `
"boolean
"
`;

exports[`parseTypeMeta > 'catchall' 1`] = `
"{
    [key: string]: Pet;
}
"
`;

exports[`parseTypeMeta > 'date' 1`] = `
"Date
"
`;

exports[`parseTypeMeta > 'datetime' 1`] = `
"string
"
`;

exports[`parseTypeMeta > 'default' 1`] = `undefined`;

exports[`parseTypeMeta > 'default' 2`] = `undefined`;

exports[`parseTypeMeta > 'enum' 1`] = `
"Enum
"
`;

exports[`parseTypeMeta > 'matches' 1`] = `undefined`;

exports[`parseTypeMeta > 'matchesReg' 1`] = `undefined`;

exports[`parseTypeMeta > 'max' 1`] = `undefined`;

exports[`parseTypeMeta > 'min' 1`] = `undefined`;

exports[`parseTypeMeta > 'nullable' 1`] = `undefined`;

exports[`parseTypeMeta > 'number' 1`] = `
"number
"
`;

exports[`parseTypeMeta > 'object' 1`] = `
"{
    firstName: string;
    /**
     * @description \\"Your address\\"
    */
    address: string | null;
}
"
`;

exports[`parseTypeMeta > 'objectAnd' 1`] = `
"{
    /**
     * @deprecated
     * @default test
    */
    firstName: string;
    /**
     * @default 2
     * @example 2
    */
    age: number;
    /**
     * @description Your address
    */
    address: (string & number) | null;
}
"
`;

exports[`parseTypeMeta > 'objectEmpty' 1`] = `
"{}
"
`;

exports[`parseTypeMeta > 'objectOptional' 1`] = `
"{
    firstName?: string;
    /**
     * @description \\"Your address\\"
    */
    address: string | null;
}
"
`;

exports[`parseTypeMeta > 'ref' 1`] = `
"Pet
"
`;

exports[`parseTypeMeta > 'string' 1`] = `
"string
"
`;

exports[`parseTypeMeta > 'tuple' 1`] = `
"[
]
"
`;

exports[`parseTypeMeta > 'tupleMulti' 1`] = `
"[
    string,
    number
]
"
`;

exports[`parseTypeMeta > 'undefined' 1`] = `
"undefined
"
`;

exports[`parseTypeMeta > 'union' 1`] = `
"(string | number)
"
`;

exports[`parseTypeMeta > 'unionOne' 1`] = `
"string
"
`;

exports[`parseTypeMeta > 'unknown' 1`] = `
"unknown
"
`;

exports[`typeParser > 'Object' 1`] = `
"export type Object = {
    /**
     * @deprecated
     * @default test
    */
    firstName: string;
    /**
     * @default 2
     * @example 2
    */
    age: number;
    /**
     * @description Your address
    */
    address: (string & number) | null;
} | null;
"
`;

exports[`typeParser > 'Order' 1`] = `
"export const orderStatus = {
    "Placed": "placed",
    "Approved": "approved"
} as const;
export type OrderStatus = (typeof orderStatus)[keyof typeof orderStatus];
export type Order = {
    status: OrderStatus;
};
"
`;

exports[`typeParser > 'Record' 1`] = `
"export type Record = {
    [key: string]: number;
};
"
`;

exports[`typeParser > 'Upload' 1`] = `
"export type Upload = Blob;
"
`;
